addons/website/models/website.py

```diff
@@ -629,125 +587,125 @@ def replace_id(old_id, new_id):
+            self.website_ids.write(data)
+        params = self.env.context.get('params', {})
         if 'url_return' in params:
             return {
-                'url': params['url_return'].replace('[lang]', language_obj['lang']),
+                'url': params['url_return'].replace('[lang]', self.lang),
                 'type': 'ir.actions.act_url',
                 'target': 'self'
             }
         return action
 
-class website_seo_metadata(osv.AbstractModel):
+class WebsiteSeoMetadata(models.AbstractModel):
     _name = 'website.seo.metadata'
     _description = 'SEO metadata'
 
-    _columns = {
-        'website_meta_title': fields.char("Website meta title", translate=True),
-        'website_meta_description': fields.text("Website meta description", translate=True),
-        'website_meta_keywords': fields.char("Website meta keywords", translate=True),
-    }
+    website_meta_title = fields.Char("Website meta title", translate=True)
+    website_meta_description = fields.Text("Website meta description", translate=True)
+    website_meta_keywords = fields.Char("Website meta keywords", translate=True)
 
 
-class website_published_mixin(osv.AbstractModel):
+class WebsitePublishedMixin(models.AbstractModel):
     _name = "website.published.mixin"
 
-    _website_url_proxy = lambda self, *a, **kw: self._website_url(*a, **kw)
+    website_published = fields.Boolean('Visible in Website', copy=False)
+    website_url = fields.Char(compute='_website_url_proxy', string='Website URL',
+                                   help='The full URL to access the document through the website.')
 
-    _columns = {
-        'website_published': fields.boolean('Visible in Website', copy=False),
-        'website_url': fields.function(_website_url_proxy, type='char', string='Website URL',
-                                       help='The full URL to access the document through the website.'),
-    }
+    # TODO remove proxy and _website_url unused arguments
+    def _website_url_proxy(self):
+        result = self._website_url(False, False)
+        for website in self:
+            website.website_url = result.get(website.id)
 
-    def _website_url(self, cr, uid, ids, field_name, arg, context=None):
-        return dict.fromkeys(ids, '#')
+    @api.multi
+    def _website_url(self, field_name, arg):
+        return dict.fromkeys(self.ids, '#')
 
-    def website_publish_button(self, cr, uid, ids, context=None):
-        for i in self.browse(cr, uid, ids, context):
-            if self.pool['res.users'].has_group(cr, uid, 'base.group_website_publisher') and i.website_url != '#':
-                return self.open_website_url(cr, uid, ids, context)
-            i.write({'website_published': not i.website_published})
-        return True
+    @api.multi
+    def website_publish_button(self):
+        self.ensure_one()
+        if self.env.user.has_group('base.group_website_publisher') and self.website_url != '#':
```

Comment:
SAME AS ABOVE
Be carefull ! The `has_group` is a `api.model` method ; there you are checking if the user that browse `request.env.user` has the group, but the user from the environment is browse as sudo.
The good code is `request.env['res.users'].has_group('base.group_portal')`